### PR TITLE
Experimental Singularity Store

### PR DESCRIPTION
**Pull request by hannahhoward**
Implement an experimental option to use Singularity as Motion blob.Store.

A singularity blob store for now is a local store + a singularity connector

For a put operation, the file is saved to the local blob store, then the local file is added to Singularity for scanning (as an item). The returned ID is the ID generated by singularity, rather than the UUID of the file (note: this changes the ID type to a raw string -- currently singularity uses integer indexes for each item -- @xinaxu maybe has an opinion about whether Singularity might either convert IDs to UUIDs or add a UUID at least for Item -- alternatively, we can make a single motion ID -> singularity primary key mapping in a file held in memory).

For a get operation, the id is sent to Singularity to retrieve metadata about the file, and then the file is retrieved from the local store. (for real retrieval with Filecoin, we could have the GetItem call to Singularity also hydrate the local store, though this may not be the most performant option)

Note that all this really does so far is start creating entities within Singularities database and scan files into Items & ItemParts.

I haven't attempted to start data prep workers or make filecoin deals.

Also, currently the singularity connector is implemented with a direct connection to an SQLite DB setup for Singularity, but could also be implemented over the singularity API.

Per the thinking in https://github.com/filecoin-project/motion/discussions/24, what may make most sense is to get #26 fully working to allow partners to try storing deals on Filecoin, while we continue to build this solution out and harden/productionize Singularity's deal making.

### CODE REVIEW

**Comment by hannahhoward** [comment]
I believe the fix for Windows tests is in @masih 's PR

File: singularity/local.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+			var source model.Source
+			err = db.
+				Where(model.Source{
+					DatasetID: dataset.ID,
+					Type:      "local",
+					Path:      dir,
+				}).Attrs(model.Source{
+				Metadata:            model.Metadata(map[string]string{}),
+				ScanIntervalSeconds: 0,
+				ScanningState:       model.Ready,
```
Comment (xinaxu) [inline singularity/local.go:86]: If it is ready, then the first time dataset worker is run, it will start scanning this directory.
If you make it completed, then the directory will never be scanned and the only way to add item is to use pushItem

**Comment by xinaxu** [comment]
> I haven't attempted to start data prep workers or make filecoin deals.

Do you plan to run data prep worker and deal making worker as separate processes, or will you run them within the same binary by invoking the underlying function?

File: singularity/local.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+type LocalSingularity struct {
+	db                        *gorm.DB
+	datasourceHandlerResolver datasource.HandlerResolver
+	initialized               bool
+	sourceID                  uint32
+}
+
+func OpenSQLiteSingularity(dbFile string) (*LocalSingularity, error) {
+	db, err := database.Open("sqlite:"+dbFile, &gorm.Config{})
```
Comment (xinaxu) [inline singularity/local.go:34]: Do you want to use this?
https://github.com/data-preservation-programs/singularity/blob/9e60f0b83c3dc29af1906402a8bd2f433b1e6780/database/connstring.go#L21
So that you can just use a single connection string for all supported backends
https://github.com/data-preservation-programs/singularity/blob/9e60f0b83c3dc29af1906402a8bd2f433b1e6780/cmd/app.go#L34

It also includes some pragmas that enables foreign key support in SQlite

File: blob/blob.go [hunk 3]
```diff
@@ -16,7 +14,7 @@ var (
 
 type (
 	// ID uniquely identifies a blob.
-	ID uuid.UUID
+	ID string
```
Comment (masih) [inline blob/blob.go:17]: I am curious what is the rationale for this change?

File: blob/local_store.go [hunk 4]
```diff
@@ -25,12 +54,20 @@ func NewLocalStore(dir string) *LocalStore {
 	}
 }
 
+func (l *LocalStore) Start(_ context.Context) error {
```
Comment (masih) [inline blob/local_store.go:57]:
```suggestion
func (l *LocalStore) Start(context.Context) error {
```

File: blob/local_store.go [hunk 5]
```diff
@@ -25,12 +54,20 @@ func NewLocalStore(dir string) *LocalStore {
...
+	return nil
+}
+
+func (l *LocalStore) Shutdown(_ context.Context) error {
```
Comment (masih) [inline blob/local_store.go:61]:
```suggestion
func (l *LocalStore) Shutdown(context.Context) error {
```

File: blob/singularity_store.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		local:       local,
+		singularity: singularity,
+	}
+}
+
+func (l *SingularityStore) Start(ctx context.Context) error {
+	return l.singularity.Start(ctx, l.local.dir)
+}
+
+func (l *SingularityStore) Shutdown(_ context.Context) error {
```
Comment (masih) [inline blob/singularity_store.go:48]:
```suggestion
func (l *SingularityStore) Shutdown(context.Context) error {
```

File: blob/singularity_store.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+func NewSingularityStore(local *LocalStore, singularity SingularityService) *SingularityStore {
+	return &SingularityStore{
+		local:       local,
+		singularity: singularity,
+	}
+}
+
+func (l *SingularityStore) Start(ctx context.Context) error {
+	return l.singularity.Start(ctx, l.local.dir)
```
Comment (masih) [inline blob/singularity_store.go:45]: Looks like this is the only functional call that changed the blobstore interface to add `Start` and `Shutdown`? My vote would be to keep the blob.Store interface as small as possible.

Overall, my plan re Start Shutdown was to introduce a service interface and type check in `Motion` to start stop store if it is a service since there will be many more such service components that `Motion` would need to start stop.

File: cmd/motion/main.go [hunk 8]
```diff
@@ -27,12 +27,29 @@ func main() {
...
 			storeDir := cctx.String("storeDir")
-			logger.Infow("Using local blob store", "storeDir", storeDir)
+			var store blob.Store
+			if cctx.Bool("experimentalSingularityStore") {
+				var err error
+				store, err = blob.OpenSingularityStore(storeDir)
+				if err != nil {
+					return err
+				}
+				logger.Infow("Using Singularity blob store", "storeDir", storeDir)
```
Comment (masih) [inline cmd/motion/main.go:45]: Add the `.Start` call here and keep the blob.Store interface small?

File: singularity/local.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	ErrSourceNotFound = errors.New("no source is found with the given id")
+	ErrNotInitialized = errors.New("singularity source not initialized")
+)
+
+const MOTION_DATASET_NAME = "MOTION_DATASET"
+const PIECE_SIZE = 1 << 36
+const MAX_CAR_SIZE = (1 << 29) * 127
+
+type LocalSingularity struct {
+	db                        *gorm.DB
```
Comment (masih) [inline singularity/local.go:27]: It seems unfortunate to have to depend on gorm just to use Singularity. 

In the long run I think Singularity should be refactored into a reusable  library and avoid exposing database models directly to the users. All sorts of funky could happen if it's exposed.

**Comment by masih** [comment]
> I believe the fix for Windows tests is in @masih 's PR

If this is referring to disabling 32bit tests, it would fix it but i don't think windows tests are failing because of that. 
Note that 32bit tests are passing on other platforms in the CI run. So the failures are specific to windows and not specific to 32b tests.

File: blob/singularity_store.go (hunk 7 above)
Comment (hannahhoward) [reply blob/singularity_store.go:45]: I did it cause of the Start/Shutdown also being in the RIBS interface. I agree though, let's keep it small.

File: cmd/motion/main.go (hunk 8 above)
Comment (hannahhoward) [reply cmd/motion/main.go:45]: Agree fine with that.

File: blob/singularity_store.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/data-preservation-programs/singularity/client"
+	"github.com/data-preservation-programs/singularity/handler/dataset"
+	"github.com/data-preservation-programs/singularity/handler/datasource"
+)
+
+const MOTION_DATASET_NAME = "MOTION_DATASET"
+const MAX_CAR_SIZE = "31.5GiB"
```
Comment (masih) [inline blob/singularity_store.go:19]: * Use camel case to stay consistent with general go naming conventions this repo follows?
* Don't export these values?

File: cmd/motion/main.go [hunk 11]
```diff
@@ -45,6 +50,16 @@ func main() {
 				Usage: "Whether to generate the local wallet key if none is found",
 				Value: true,
 			},
+			&cli.BoolFlag{
+				Name:        "experimentalSingularityStore",
+				Usage:       "whether to use experimental Singularity store as the storage and deal making engine",
+				DefaultText: "Local storage is used",
+			},
+			&cli.StringFlag{
+				Name:        "remoteSingularityAPIUrl",
```
Comment (masih) [inline cmd/motion/main.go:59]: Add `experimental` prefix?

File: cmd/motion/main.go [hunk 12]
```diff
@@ -76,6 +91,30 @@ func main() {
 					return err
 				}
 				store = rbstore
+			} else if cctx.Bool("experimentalSingularityStore") {
+				singularityAPIUrl := cctx.String("remoteSingularityAPIUrl")
+				var client client.Client
+				if singularityAPIUrl != "" {
+					client = httpclient.NewHTTPClient(http.DefaultClient, singularityAPIUrl)
+				} else {
+					db, err := database.OpenWithDefaults("sqlite:" + storeDir + "/singularity.db")
```
Comment (masih) [inline cmd/motion/main.go:100]: I am curious what would the benefit of supporting this? 

For the Singularity options to work, singularity needs to be running anyway and running singularity API to then only communicate remotely seems like a that that is much less error prone and would give Singularity full ownership of its own database.

File: cmd/motion/main.go [hunk 13]
```diff
@@ -76,6 +91,30 @@ func main() {
 					return err
 				}
 				store = rbstore
+			} else if cctx.Bool("experimentalSingularityStore") {
+				singularityAPIUrl := cctx.String("remoteSingularityAPIUrl")
+				var client client.Client
+				if singularityAPIUrl != "" {
+					client = httpclient.NewHTTPClient(http.DefaultClient, singularityAPIUrl)
```
Comment (masih) [inline cmd/motion/main.go:98]: I left [a question](https://github.com/data-preservation-programs/singularity/pull/94#issuecomment-1653994236) on Singularity repo about the long term vision for this HTTP client. 

If it's not going to be a 1:1 API to the HTTP API then would it make sense to call this something else considering there is now an auto-generated HTTP client library that fully covers the Singularity API?

File: blob/singularity_store.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+	var asDuplicatedRecord client.DuplicateRecordError
+
+	// return errors, but ignore duplicated record, that means we just already created it
+	if err != nil && !errors.As(err, &asDuplicatedRecord) {
+		return err
+	}
+	source, err := l.singularityClient.CreateLocalSource(ctx, MOTION_DATASET_NAME, datasource.LocalRequest{
+		SourcePath:        l.local.dir,
+		RescanInterval:    "0",
+		DeleteAfterExport: false,
```
Comment (masih) [inline blob/singularity_store.go:49]: This means stored files will remain in local directory indefinietly. Correct?

**Review by masih — COMMENTED** [review]
LGTM other than the failing go check CI

Left some questions.

File: blob/singularity_store.go (hunk 14 above)
Comment (hannahhoward) [reply blob/singularity_store.go:49]: for now yes.

Comment (hannahhoward) [reply blob/singularity_store.go:49]: we will delete once retrieval is available. though possibly not through this flag.
